qemu/tests/watchdog.py

```diff
@@ -89,12 +89,40 @@ def _action_check(test, session, watchdog_action):
         """
         # when watchdog action is pause, shutdown, reset, poweroff
         # the vm session will lost responsive
+
+        def check_guest_reboot(pattern):
+            start_time = time.time()
+            while (time.time() - start_time) < vm.REBOOT_TIMEOUT:
+                if pattern in vm.serial_console.get_output():
+                    return True
+            return False
+
         response_timeout = int(params.get("response_timeout", '240'))
+        vm_arch_name = params["vm_arch_name"]
         error_context.context("Check whether or not watchdog action '%s' took"
                               " effect" % watchdog_action, logging.info)
+        if watchdog_action == "inject-nmi":
+            vm.monitor.clear_event("WATCHDOG")
+            if (vm_arch_name in ("x86_64", "i686")):
+                if not utils_misc.wait_for(lambda: vm.monitor.get_event("WATCHDOG"),
+                                           timeout=response_timeout):
+                    test.fail("Watchodg action '%s' didn't take effect."
+                              % watchdog_action)
```

Comment:
No, this is only for inject-nmi on x86. Because for other cases, there will be specific results to check if the case passes, no need to check monitor event. But for inject-nmi on x86, no other results except for monitor event, guest will not reboot, pause ..., so I check if the inject-nmi case of x86 passes via monitor event.

Reply:
Well, there are several ways to check it.

1. kernel message (dmesg)
```
[  121.564648] Uhhuh. NMI received for unknown reason 20 on CPU 0.
```
2. interrupts table
```
[root@test ~]# cat /proc/interrupts | grep NMI
NMI:          0          0          0          0   Non-maskable interrupts
[root@test ~]# 
Message from syslogd@test at Jan 25 12:52:07 ...
 kernel:Uhhuh. NMI received for unknown reason 20 on CPU 0.

Message from syslogd@test at Jan 25 12:52:07 ...
 kernel:Do you have a strange power saving mode enabled?

Message from syslogd@test at Jan 25 12:52:07 ...
 kernel:Dazed and confused, but trying to continue

[root@test ~]# cat /proc/interrupts | grep NMI
NMI:          1          0          0          0   Non-maskable interrupts
```

not suggest only making the check work for x86 machines, because:
1. it does not work at all when using hmp
2. other tests also support it

Reply:
@luckyh I have added check for dmesg info. And for vm.monitor, I have specified  monitor_type = qmp.
x86_64:
 (1/2) Host_RHEL.m8.u0.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.6.x86_64.myprovider.watchdog.watchdog_action.inject_nmi.i6300esb: PASS (317.44 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.6.x86_64.myprovider.watchdog.watchdog_action.inject_nmi.ib700: PASS (659.03 s)
power:
 (1/1) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.watchdog.watchdog_action.inject_nmi.i6300esb: PASS (405.14 s)

s390x:
No machine now, I will test it later.

Reply:
@luckyh For s390x, there is function bug now, guest will be panic when running the case, and the case for x86  is important,  so now I will only support x86 and power. Could you help review again?